subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java

```diff
@@ -87,14 +91,37 @@ public void run() {
 
     @Override
     public boolean isCacheEnabled() {
-        return !cacheIfSpec.getSpecs().isEmpty() && cacheIfSpec.isSatisfiedBy(task)
-            && (doNotCacheIfSpec.isEmpty() || !doNotCacheIfSpec.isSatisfiedBy(task));
+        return isCacheEnabled(new ArrayList<TaskCachingDisabledReason>());
     }
 
     @Override
-    public boolean isCacheAllowed() {
+    public boolean isCacheEnabled(List<TaskCachingDisabledReason> reasons) {
```

Comment:
Currently, we are not re-evaluating the spec multiple times - we are only calling it from the `SkipCachedTaskExecuter`. The result of evaluating the spec could change over time and anyone can call this method via `task.outputs.taskCaching`.
Should we add some validation and only execute the method once or leave it as is?